### DISCUSSION TITLE
Connect time for python/bleak on Windows is slow compared to iOS native applications

### DISCUSSION THREAD

**Discussion by sigsbeeus** [hbldh/bleak#910] [Q&A]
I use python and bleak on WIndows for automated BLE testing for a BLE device under development.  A native iOS application is used to communicate with the device for customers.   Our device goes into a low power advertising mode with advertisements every 3s after several minutes of inactivity.  The iOS application has no problem connecting the the device quickly, often within a few seconds.  Windows + python + Bleak can take many tens of seconds to connect, and sometimes cannot connect at all.  Is there a special connection parameter or request that can be made from Bleak to do whatever iOS is doing?   There must be something iOS is doing to improve connection time for a slowly advertising device, but i cannot find out what that is anywhere.

Thank you,
--b

**Comment by dlech** [comment]
>  Is there a special connection parameter or request that can be made from Bleak to do whatever iOS is doing? 

If there is one, we would sure like to know about it. But it seems that Windows has just implemented their Bluetooth stack very differently compared to other OSes.

I would suggest logging Bluetooth packets using Wireshark (see troubleshooting page of bleak docs) to see what Windows is actually doing under the hood and compare it to a capture on another OS.

**Reply by cmjonze** [reply]
Looking at wireshark captures using three ubertooth boards, the answer is it takes MUCH longer for the CONNECT_IND packet to go out using bleak and windows.   This is evident even from the timestamps on the log messages:

2022-08-01 08:10:16,956 [DEBUG] bleak.backends.winrt.client: Connecting to BLE device @ e0:c0:bf:18:4e:ff
2022-08-01 08:10:56,279 [DEBUG] bleak.backends.winrt.client: _ConnectionStatusChanged_Handler: 1
2022-08-01 08:10:56,281 [DEBUG] bleak.backends.winrt.client: Get Services...

The delay is definitely a function of the advertising interval of the device being connected to.  A long advertising interval, greater than a second, can cause extreme delays connecting.   Does anyone have a clue about what's going on, or is there some parameter i can set somewhere in the windows stack to help deal with this?

It would be interesting to be able to find a windows-equivalent of some BLE utility like the LightBlue app from PunchThrough.  A utilility that will let you connect to a particular device to look at characteristics.   I've had difficultiy finding something like that in the windows world.   Does anyone have a suggestion?   I'd be interested to see if a straight-up windows utility has the same issue that i'm seeing with python.

**Reply by dlech** [reply]
> _ConnectionStatusChanged_Handler

Which version of Bleak are you using? There is currently nothing in Bleak that would print this message.

**Reply by dlech** [reply]
> It would be interesting to be able to find a windows-equivalent of some BLE utility like the LightBlue app from PunchThrough. A utilility that will let you connect to a particular device to look at characteristics. 

A quick and easy solution could be to use `chrome://bluetooth-internals/#devices` in a Chromium-based web browser.

There is also GitHub: https://github.com/Microsoft/BluetoothLEExplorer which is available in the Windows Store.

**Reply by cmjonze** [reply]
Thanks for the chrome solution, didn't know about that.   I'd also forgotten that we'd added some logging to try to figure out what's happening.  I do see the same behavior in the latest, unadulterated version of bleak in git.   Chrome seems to also have trouble digging the device out of its long advertising interval to do a connect.  I need to run some more tests to tell if there's really a difference between what's going on in python/bleak and the chrome thing.

I guess I shouldn't be surprised that windows turns out to be kind of lame again.  The device we're working on has very good reasons to minimize radio traffic as much as it can because it's on a small battery that needs to last a long time.   It's weird that windows seems to be so much worse at connecting than iOS.   Have not tried android, by the way.

Thanks again for the info.  If I can figure out anything definite I'll post back.
